aggregator/src/aggregator/aggregation_job_driver.rs

```diff
@@ -772,16 +898,155 @@ where
+                    stepped_aggregations,
+                    report_aggregations_to_write,
+                    prepare_resps,
+                )
+                .await
+            }
+        }
+    }
+
+    async fn process_response_from_helper_pending<
+        const SEED_SIZE: usize,
+        C: Clock,
+        B: CollectableBatchMode,
+        A: vdaf::Aggregator<SEED_SIZE, 16> + Send + Sync + 'static,
+    >(
+        &self,
+        datastore: Arc<Datastore<C>>,
+        vdaf: Arc<A>,
+        lease: Arc<Lease<AcquiredAggregationJob>>,
+        task: Arc<AggregatorTask>,
+        aggregation_job: AggregationJob<SEED_SIZE, B, A>,
+        stepped_aggregations: Vec<SteppedAggregation<SEED_SIZE, A>>,
+        mut report_aggregations_to_write: Vec<WritableReportAggregation<SEED_SIZE, A>>,
+        retry_after: Option<&Duration>,
+    ) -> Result<(), Error>
+    where
+        A::AggregationParam: Send + Sync + Eq + PartialEq,
+        A::AggregateShare: Send + Sync,
+        A::InputShare: Send + Sync,
+        A::OutputShare: Send + Sync,
+        A::PrepareMessage: Send + Sync,
+        A::PrepareShare: Send + Sync,
+        A::PrepareState: Send + Sync + Encode,
+        A::PublicShare: Send + Sync,
+    {
+        // Any non-failed report aggregations are set to the Poll state, allowing them to be polled
+        // when the aggregation job is next picked up.
+        report_aggregations_to_write.extend(stepped_aggregations.into_iter().map(
+            |stepped_aggregation| {
+                WritableReportAggregation::new(
+                    stepped_aggregation.report_aggregation.with_state(
+                        ReportAggregationState::LeaderPoll {
+                            leader_state: stepped_aggregation.leader_state,
+                        },
+                    ),
+                    // Even if we have recovered an output share (i.e.,
+                    // `stepped_aggregation.leader_state` is Finished), we don't include it here: we
+                    // aren't done with aggregation until we receive a response from the Helper, so
+                    // it would be incorrect to merge the results into the batch aggregations at
+                    // this point.
+                    None,
+                )
+            },
+        ));
+
+        // Write everything back to storage.
+        let mut aggregation_job_writer =
+            AggregationJobWriter::<SEED_SIZE, _, _, UpdateWrite, _>::new(
+                Arc::clone(&task),
+                self.batch_aggregation_shard_count,
+                Some(AggregationJobWriterMetrics {
+                    report_aggregation_success_counter: self.aggregation_success_counter.clone(),
+                    aggregate_step_failure_counter: self.aggregate_step_failure_counter.clone(),
+                    aggregated_report_share_dimension_histogram: self
+                        .aggregated_report_share_dimension_histogram
+                        .clone(),
```

Comment:
We should rethink Prometheus metrics in the context of `AggregationJobResp::Processing` responses as well. I haven't looked at the metrics in detail yet, but we may want additional counters, or new labels, in order to distinguish forward progress versus mere polling.

Reply:
I think this is wise, but these metrics only concern themselves with completed report aggregations & failed steps, both of which are agnostic to the aggregation mode. I think we can get pretty far by looking to the request verb & the response code, though full these metrics won't give full clarity (i.e. we can differentiate between a Leader's attempt to poll vs an attempt to send a new step via the HTTP verb; but I don't think we can tell whether the Helper responded with a `processing` or `finished` response with these metrics). However, I strongly suspect that our current integrations will always use exactly one of asynchronous or synchronous responses, so this may be moot at the moment, at least for us.